plugins/inputs/huawei_routers_telemetry/huawei_routers_telemetry.go

```diff
@@ -0,0 +1,219 @@
+package huawei_routers_telemetry
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/DamRCorba/huawei_telemetry_sensors"
+	"github.com/DamRCorba/huawei_telemetry_sensors/sensors/huawei-telemetry"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type setReadBufferer interface {
+	SetReadBuffer(bytes int) error
+}
+
+type streamSocketListener struct {
+	net.Listener
+	*HuaweiRoutersTelemetry
+
+	sockType string
+
+	connections    map[string]net.Conn
+	connectionsMtx sync.Mutex
+}
+
+type packetSocketListener struct {
+	net.PacketConn
+	*HuaweiRoutersTelemetry
+}
+
+/*
+  Telemetry Decoder.
+
+*/
+func HuaweiTelemetryDecoder(body []byte) (*metric.SeriesGrouper, error) {
+	msg := &telemetry.Telemetry{}
+	err := proto.Unmarshal(body[12:], msg)
+	if err != nil {
+		fmt.Println("Unable to decode incoming packet: ", err.Error())
+		return nil, err
+		//panic(err)
+	}
+	grouper := metric.NewSeriesGrouper()
+	for _, gpbkv := range msg.GetDataGpb().GetRow() {
+		dataTime := gpbkv.Timestamp
+		if dataTime == 0 {
+			dataTime = msg.MsgTimestamp
+		}
+		timestamp := time.Unix(int64(dataTime/1000), int64(dataTime%1000)*1000000)
+		sensorMsg := huawei_sensorPath.GetMessageType(msg.GetSensorPath())
+		err = proto.Unmarshal(gpbkv.Content, sensorMsg)
+		if err != nil {
+			fmt.Println("Sensor Error: ", err.Error())
+			return nil, err
+		}
+		fields, vals := huawei_sensorPath.SearchKey(gpbkv, msg.GetSensorPath())
+		tags := make(map[string]string, len(fields)+3)
+		tags["source"] = msg.GetNodeIdStr()
+		tags["subscription"] = msg.GetSubscriptionIdStr()
+		tags["path"] = msg.GetSensorPath()
+		// Search for Tags
+		for i := 0; i < len(fields); i++ {
+			tags = huawei_sensorPath.AppendTags(fields[i], vals[i], tags, msg.GetSensorPath())
+		}
+		// Create Metrics
+		for i := 0; i < len(fields); i++ {
+			huawei_sensorPath.CreateMetrics(grouper, tags, timestamp, msg.GetSensorPath(), fields[i], vals[i])
```

Comment:
No. Please do not do this in the external package! First of all you add a dependency to _telegraf_ in your package! While this is not the problem of telegraf, you create a big ball of mud with intertwined code totally killing modularization! Furthermore, nobody will be able to debug problems without modifying your packages as you cannot get any idea of what data is involved.

Please modify the `huawei_sensorPath.CreateMetrics()` to return the data in a sensible way _without_ relying on the grouper and then group the data here.

Reply:
In progress